src/lib.rs

```diff
@@ -250,31 +247,86 @@ impl<TSuccess> From<NonFatalSearchError> for Result<TSuccess, SingleFileSearchEr
     }
 }
 
-pub fn run(args: Args) -> Result<RunStatus, Error> {
-    let query_text = match (args.path_to_query_file.as_ref(), args.query_text.as_ref()) {
-        (Some(path_to_query_file), None) => {
-            fs::read_to_string(path_to_query_file).map_err(|source| Error::QueryFileReadError {
-                source,
-                path_to_query_file: path_to_query_file.clone(),
-            })?
-        }
-        (None, Some(query_text)) => query_text.clone(),
-        (None, None) => ALL_NODES_QUERY.to_owned(),
-        _ => unreachable!(),
-    };
-    let filter =
-        get_loaded_filter(args.filter.as_deref(), args.filter_arg.as_deref())?.map(Arc::new);
+pub struct OutputContext {
+    pub buffer_writer: BufferWriter,
+}
+
+impl OutputContext {
+    pub fn new(buffer_writer: BufferWriter) -> Self {
+        Self { buffer_writer }
+    }
+}
+
+pub fn run_print(args: Args) -> Result<RunStatus, Error> {
+    run_for_context(
+        args,
+        OutputContext::new(BufferWriter::stdout(ColorChoice::Never)),
+        |context: &OutputContext,
```

Comment:
So then the per-file "do the actual core searching stuff" callback gets access to the stuff it needs including that context object